doc/man/man2/read.2.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the section entitled "GNU Free
.\" Documentation License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaning to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF
.\" SUCH CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log$
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore T
.R2
.so openss7.macros
.\" =========================================================================
.\"
.\" Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
.\"
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Library General Public
.\" License as published by the Free Software Foundation; either
.\" version 2 of the License, or (at your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this library; if not, write to the
.\" Free Software Foundation, Inc., 59 Temple Place - Suite 330, Cambridge,
.\" MA 02139, USA.
.\"
.\" =========================================================================
.\"
.\"
.\" Hey Emacs! This file is -*- nroff -*- source.
.\"
.\" This manpage is Copyright (C) 1992 Drew Eckhardt;
.\"                               1993 Michael Haardt, Ian Jackson.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Modified Sat Jul 24 00:06:00 1993 by Rik Faith <faith@cs.unc.edu>
.\" Modified Wed Jan 17 16:02:32 1996 by Michael Haardt
.\"   <michael@cantor.informatik.rwth-aachen.de>
.\" Modified Thu Apr 11 19:26:35 1996 by Andries Brouwer <aeb@cwi.nl>
.\" Modified Sun Jul 21 18:59:33 1996 by Andries Brouwer <aeb@cwi.nl>
.\" Modified Fri Jan 31 16:47:33 1997 by Eric S. Raymond <esr@thyrsus.com>
.\" Modified Sat Jul 12 20:45:39 1997 by Michael Haardt
.\"   <michael@cantor.informatik.rwth-aachen.de>
.\"
.TH READ 2 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Calls"
.SH NAME
.B read
\- read from a file descriptor
.SH SYNOPSIS
.nf
.B #include <unistd.h>
.sp
.BI "ssize_t read(int " fd ", void *" buf ", size_t " count );
.fi
.SH DESCRIPTION
.B read()
attempts to read up to
.I count
bytes from file descriptor
.I fd
into the buffer starting at
.IR buf .
.PP
If
.I count
is zero, \fBread()\fP returns zero and has no other results.
If
.I count
is greater than SSIZE_MAX, the result is unspecified.
.PP
.SH "RETURN VALUE"
On success, the number of bytes read is returned (zero indicates end of
file), and the file position is advanced by this number.
It is not an error if this number is smaller than the number of bytes
requested; this may happen for example because fewer bytes are actually
available right now (maybe because we were close to end-of-file, or
because we are reading from a pipe, or from a terminal), or because
\fBread()\fP was interrupted by a signal.
On error, \-1 is returned, and
.I errno
is set appropriately. In this case it is left unspecified whether
the file position (if any) changes.
.SH ERRORS
.TP
.B EINTR
The call was interrupted by a signal before any data was read.
.TP
.B EAGAIN 
Non-blocking I/O has been selected using
.B O_NONBLOCK
and no data was immediately available for reading.
.TP
.B EIO
I/O error. This will happen for example when the process is in a
background process group, tries to read from its controlling tty,
and either it is ignoring or blocking SIGTTIN or its process group
is orphaned.  It may also occur when there is a low-level I/O error
while reading from a disk or tape.
.TP
.B EISDIR
.I fd
refers to a directory.
.TP
.B EBADF
.I fd
is not a valid file descriptor or is not open for reading.
.TP
.B EINVAL
.I fd
is attached to an object which is unsuitable for reading.
.TP
.B EFAULT
.I buf
is outside your accessible address space.
.PP
Other errors may occur, depending on the object connected to
.IR fd .
POSIX allows a
.B read
that is interrupted after reading some data
to return \-1 (with
.I errno
set to EINTR) or to return the number of bytes already read.
.SH "CONFORMING TO"
SVr4, SVID, AT&T, POSIX, X/OPEN, BSD 4.3
.SH RESTRICTIONS
On NFS file systems, reading small amounts of data will only update the
time stamp the first time, subsequent calls may not do so.  This is caused
by client side attribute caching, because most if not all NFS clients
leave atime updates to the server and client side reads satisfied from the
client's cache will not cause atime updates on the server as there are no
server side reads.  UNIX semantics can be obtained by disabling client
side attribute caching, but in most situations this will substantially
increase server load and decrease performance.
.SH "SEE ALSO"
.BR close (2),
.BR fcntl (2),
.BR ioctl (2s),
.BR lseek (2),
.BR readdir (2),
.BR readlink (2),
.BR select (2),
.BR write (2s),
.BR fread (3)
.[
$LIST$
.]


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-04-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel20081029_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.

   - Modifications to build under `Fedora' `2.6.22.5-49' kernel.  These
     changes also support `2.6.22.9-91.fc7' kernel.  Modifications to
     build under `Fedora' `2.6.25-45.fc9' and `2.6.26.5-45.fc9' kernels.
     Documented `lib32gcc1' problem on Ubuntu.  Noted problem running
     under XEN kernels.  XEN kernels are not yet supported.  Added
     `MODULE_VERSION' to all modules and drivers.

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.  Improvements to common build
     process for embedded and cross-compile targets.  Cross-compile
     fixes (strap out `AC_FUNC_REALLOC' macro when cross-compiling).
     Conversion of RPM spec files to common approach for major
     subpackages.

     Build system now builds `yum(8)' repositories for RPMs and
     `apt-get(8)' repositories for DEBs.  Installation documentation
     has been updated to include details of repository install
     sourcesref.

   - Higher performance and updated performance papers.

   - Updated tool chain to `m4-1.4.12', `autoconf-2.63' and
     `texinfo-4.13'.  Support for `flex 2.5.33' in maintainer mode.

   - Updated references database for manual pages and roff documents.

   - Added the following major sub-packages to the master build (and
     release):


isode/others/quipu/uips/pod/read.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

#include "quipu/util.h"
#include "quipu/common.h"
#include "quipu/entry.h"
#include "quipu/name.h"

#include <sys/types.h>
#include <ctype.h>

#include "sequence.h"
#include "defs.h"
#include "util.h"

#ifndef NO_STATS
extern LLog *log_stat;
#endif

int as_print();
void kill_photo();
void readEntryPrint();

extern char goto_path[], base_path[], friendly_base_path[];
extern int dn_number, two_passes;

mailtype mailformat = rfc822;
void rfc2greybook();

Attr_Sequence read_types = 0, read_types2 = 0, oclass = 0;
bool photo_on = TRUE;

dsEnqError do_read(), read_all(), read_config_types();

Attr_Sequence
sort_attrs(entry_attrs)
	Attr_Sequence entry_attrs;
{
	register Attr_Sequence last, next, curr, first, firstn;

	first = curr = entry_attrs;
	firstn = last = next = 0;

	while (curr)
		if (!strcmp("2.5.4.3", curr->attr_type->oa_ot.ot_stroid) ||
		    !strcmp("2.5.4.4", curr->attr_type->oa_ot.ot_stroid) ||
		    !strcmp("0.9.2342.19200300.100.1.3",
			    curr->attr_type->oa_ot.ot_stroid) ||
		    !strcmp("0.9.2342.19200300.100.1.2",
			    curr->attr_type->oa_ot.ot_stroid) ||
		    !strcmp("2.5.4.20", curr->attr_type->oa_ot.ot_stroid)) {

			if (first == curr)
				first = curr->attr_link;

			if (next)
				next->attr_link = curr;
			else
				firstn = curr;

			next = curr;

			if (last)
				last->attr_link = curr->attr_link;

			curr = curr->attr_link;
			next->attr_link = 0;
		} else {
			last = curr;
			curr = curr->attr_link;
		}

	if (next) {
		next->attr_link = first;
		return firstn;
	} else
		return first;
}

Attr_Sequence
get_sorted_attrs(entry_attrs, wanted_attrs)
	Attr_Sequence entry_attrs, wanted_attrs;
{
	register Attr_Sequence curr_ent_attr, curr_found, curr_wanted, attrs_found, next;

	attrs_found = NULLATTR;
	curr_wanted = wanted_attrs;
	curr_ent_attr = entry_attrs;

	while (curr_wanted != NULLATTR) {
		while (curr_ent_attr != NULLATTR) {
			if (AttrT_cmp(curr_wanted->attr_type, curr_ent_attr->attr_type) == 0) {
				next = curr_ent_attr->attr_link;
				curr_ent_attr->attr_link = NULLATTR;

				if (attrs_found == NULLATTR) {
					curr_found = attrs_found = as_cpy(curr_ent_attr);
				} else {
					curr_found->attr_link = as_cpy(curr_ent_attr);
					curr_found = curr_found->attr_link;
				}

				curr_ent_attr->attr_link = next;
				curr_ent_attr = next;
			} else
				curr_ent_attr = curr_ent_attr->attr_link;
		}
		curr_wanted = curr_wanted->attr_link;
		curr_ent_attr = entry_attrs;
	}

	curr_wanted = attrs_found;

	attrs_found = sort_attrs(attrs_found);
	return attrs_found;
}

dsEnqError
read_all()
{
	return do_read(NULLATTR);
}

dsEnqError
read_config_types()
{
	return do_read(read_types);
}

dsEnqError
do_read(attrs_to_read)
	Attr_Sequence attrs_to_read;
{
	struct ds_read_arg read_arg;
	struct ds_read_result result;
	struct DSError error;
	Entry read_entry;
	Attr_Sequence wanted_attrs;
	extern void setReadEntryName();

	if (*base_path == '\0') {
		kill_photo();
		setReadEntryName(base_path);
		readEntryPrint("description - Big and Round (but getting smaller).\n");
		return Okay;
	}

	if (get_default_service(&read_arg.rda_common) != 0) {
		return nothingfound;
	}

	read_arg.rda_common.ca_servicecontrol.svc_options = SVC_OPT_PREFERCHAIN;

	read_arg.rda_eis.eis_allattributes = TRUE;
	read_arg.rda_eis.eis_infotypes = EIS_ATTRIBUTESANDVALUES;
	read_arg.rda_eis.eis_select = NULLATTR;

	read_arg.rda_object = (strncmp(friendly_base_path, "The World", 9) ?
			       str2dn(base_path) : NULLDN);

	if ((read_entry = local_find_entry(read_arg.rda_object, FALSE))
	    != NULLENTRY && read_entry->e_data != E_TYPE_CONSTRUCTOR) {
		kill_photo();
		setReadEntryName(base_path);

		if (attrs_to_read != NULLATTR)
			wanted_attrs = get_sorted_attrs(read_entry->e_attributes, attrs_to_read);
		else
			wanted_attrs = read_entry->e_attributes;

		read_print(as_print, (caddr_t) wanted_attrs);
		dn_free(read_arg.rda_object);

		if (wanted_attrs != read_entry->e_attributes)
			as_free(wanted_attrs);

		return Okay;
	}
#ifndef NO_STATS
	LLOG(log_stat, LLOG_NOTICE, ("read +%s", base_path));
#endif

	if (ds_read(&read_arg, &error, &result) != DS_OK) {
		int errcode;

		/* deal with error */
		log_ds_error(&error);

		errcode = error.dse_type;

		dn_free(read_arg.rda_object);
		ds_error_free(&error);

		switch (errcode) {
		case DSE_LOCALERROR:
			return duaerror;

		case DSE_REMOTEERROR:
			return localdsaerror;

		case DSE_ATTRIBUTEERROR:
			return attributerror;

		case DSE_REFERRAL:
		case DSE_DSAREFERRAL:
			return remotedsaerror;

		case DSE_SECURITYERROR:
			return security;

		case DSE_NAMEERROR:
			return namerror;

		case DSE_SERVICEERROR:
			return serviceerror;

		default:
			return localdsaerror;
		}
	} else {
		/* use data */
		if (result.rdr_entry.ent_attr == NULLATTR) {
			dn_number = 0;
			return nothingfound;
		} else {
			kill_photo();
			cache_entry(&(result.rdr_entry), TRUE, TRUE);

			if (attrs_to_read != NULLATTR)
				wanted_attrs =
				    get_sorted_attrs(result.rdr_entry.ent_attr, attrs_to_read);
			else
				wanted_attrs = result.rdr_entry.ent_attr;

			setReadEntryName(base_path);

			read_print(as_print, (caddr_t) wanted_attrs);

			if (wanted_attrs != result.rdr_entry.ent_attr)
				as_free(wanted_attrs);

			entryinfo_comp_free(&result.rdr_entry, 0);

			dn_free(read_arg.rda_object);
			return Okay;
		}
	}
}

read_print(func, ptr)
	int (*func) ();
	caddr_t ptr;
{
	PS ps;
	register char *str;

	if ((ps = ps_alloc(str_open)) == NULLPS)
		return;
	if (str_setup(ps, NULLCP, 0, 0) == NOTOK)
		return;

	(*func) (ps, ptr, READOUT);
	*--ps->ps_ptr = NULL, ps->ps_cnt++;

	str = ps->ps_base;

	ps->ps_base = NULL;
	ps->ps_cnt = 0;
	ps->ps_ptr = NULL;
	ps->ps_bufsiz = 0;

	ps_free(ps);

	readEntryPrint(str);
	free(str);
}

 /*ARGSUSED*/
podphoto(ps, picture, format)
	PS ps;
	PE picture;
	int format;
{
	PS sps;

	if (photo_on == TRUE) {
		if ((sps = ps_alloc(str_open)) == NULLPS)
			return;
		if (str_setup(sps, NULLCP, LINESIZE, 0) == NOTOK) {
			ps_free(sps);
			return;
		}

		two_passes = 0;

		(void) pe2ps(sps, picture);
		if (decode_t4(sps->ps_base, "photo", sps->ps_ptr - sps->ps_base) == -1)
			goto out;

		if (two_passes &&
		    decode_t4(sps->ps_base, "photo", sps->ps_ptr - sps->ps_base) == -1)
			goto out;

		ps_print(ps, "(see below)");

	      out:;
		ps_free(sps);
	}
}

entry2str(ptr, cptr, size)
	caddr_t ptr;
	char *cptr;
	int size;
{
	PS ps;

	if ((ps = ps_alloc(str_open)) == NULLPS)
		return;
	if (str_setup(ps, cptr, size, 1) == NOTOK)
		return;

	as_print(ps, (Attr_Sequence) ptr, READOUT);
	*ps->ps_ptr = 0;

	ps_free(ps);
}

void
rfc2greybook(string)
	char *string;
{
	char reversed[STRINGLEN];
	register char *part;

	reversed[0] = '\0';
	part = string + strlen(string);

	if (*part != '\0')
		return;

	while (1) {
		while (*part != '.' && *part != '@')
			--part;

		if (*part == '.') {
			if (reversed[0] != '\0')
				(void) strcat(reversed, ".");
			part++;
			(void) strcat(reversed, part);
			*--part = '\0';
			--part;
		} else {
			part++;
			(void) strcat(reversed, ".");
			(void) strcat(reversed, part);
			*part = '\0';
/*       while (!isspace(*part)) --part; */
			(void) strcat(string, reversed);
			return;
		}
	}
}
